docs/architecture/adr-049-state-sync-hooks.md

```diff
@@ -0,0 +1,187 @@
+calling `RegisterExtensions` when setting up the application. And the snapshotters will handle both taking snapshot and restoration.
+
+```go
+// RegisterExtensions register extension snapshotters to manager
+func (m *Manager) RegisterExtensions(extensions ...types.ExtensionSnapshotter) error 
+```
+
+Remain `Snapshotter` interface for `multistore` snapshotter. No need for `SnapshotFormat` and `SupportedFormats` as for `multistore` snapshotter
+formats are handled in a higher level.
+<ul>
+<li> removal of format parameter: `Snapshotter` chooses a format autonomously, not pass in from the caller. </li>
+<ul>
+
+```go
+// Snapshotter is something that can create and restore snapshots, consisting of streamed binary
+// chunks - all of which must be read from the channel and closed. If an unsupported format is
+// given, it must return ErrUnknownFormat (possibly wrapped with fmt.Errorf).
+type Snapshotter interface {
+	// Snapshot writes snapshot items into the protobuf writer.
+	Snapshot(height uint64, protoWriter protoio.Writer) error
+
+	// Restore restores a state snapshot from the protobuf items read from the reader.
+	// If the ready channel is non-nil, it returns a ready signal (by being closed) once the
+	// restorer is ready to accept chunks.
+	Restore(height uint64, format uint32, protoReader protoio.Reader) (SnapshotItem, error)
+}
+```
+
+Add `ExtensionSnapshotter` interface for extension snapshotters, and three more function signatures: `SnapshotFormat()` `SupportedFormats()` and
+`SnapshotName()` are added.
+
+```go
+// ExtensionSnapshotter is an extension Snapshotter that is appended to the snapshot stream.
+// ExtensionSnapshotter has an unique name and manages it's own internal formats.
+type ExtensionSnapshotter interface {
+	Snapshotter
+
+	// SnapshotName returns the name of snapshotter, it should be unique in the manager.
+	SnapshotName() string
+
+	// SnapshotFormat returns the default format used to take a snapshot.
+	SnapshotFormat() uint32
+
+	// SupportedFormats returns a list of formats it can restore from.
+	SupportedFormats() []uint32
+}
+```
+
+## Consequences
+
+As a result of this implementation, we are able to create snapshots of binary chunk stream for the state that we maintain outside of the IAVL Tree, CosmWasm blobs for example. And new clients are able to fetch sanpshots of state for all modules that have implemented the corresponding interface from peer nodes. 
+
+
+### Backwards Compatibility
+
+This ADR introduces new proto message types, add field for extension snapshotters to snapshot `Manager`, add new function signatures `SnapshotFormat()`, `SupportedFormats()` to `Snapshotter` interface and add new `ExtensionSnapshotter` interface, so this is not backwards compatible.
+
+### Positive
+
+State maintained outside of IAVL tree like CosmWasm blobs can create snapshots by implementing extension snapshotters, and being fetched by new clients via state-sync.
+
+### Negative
+
+// Todo
```

Comment:
If there are no negatives, which is entirely possible, but unlikely. I would either remove the TODO or note them.

Reply:
currently negative consequences not found.
would appreciate if anyone can inform of me about this

Reply:
lets leave it blank. The only thing I can think of is increased bandwidth but this isn't really an issue